Enable mypy checking on V1 code

markmc opened the pull request:
As per #10959

Some of these definitely deserve closer scrutiny e.g.:

- ~~commit 7974f1d - was there a bug with the `hash_block_tokens()` call?~~
- ~~commit 226fd552089ea8f0aeb49c8c2a4cc62f2b32742a - feels we'll be fighting an uphill battle with mypy and this args/kwargs pattern~~
- ~~commit 8a6d88c42df22d9ef8d1deb444a9335117b91dc6 - `MultiprocExecutor.workers` and `UniprocExecutor.worker` can be set to `None`, so we have to litter the code with `is not None` assertions~~

WoosukKwon committed:
[V1] Run mypy on

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

markmc committed:
v1: fix various mypy issues with MultiprocExecutor

Fixes:

vllm/v1/executor/abstract.py:47: error: Bracketed expression "[...]" is not valid as a type  [valid-type]
vllm/v1/executor/multiproc_executor.py:106: error: Bracketed expression "[...]" is not valid as a type  [valid-type]
vllm/v1/executor/multiproc_executor.py:128: error: "float" not callable  [operator]
vllm/v1/executor/multiproc_executor.py:175: error: Incompatible types in assignment (expression has type "None", variable has type "list[WorkerProcHandle]")  [assignment]
vllm/v1/executor/uniproc_executor.py:78: error: Incompatible types in assignment (expression has type "None", variable has type "Worker")  [assignment]
vllm/v1/engine/llm_engine.py:115: error: Incompatible types in assignment (expression has type "type[UniprocExecutor]", variable has type "type[MultiprocExecutor]")  [assignment]
vllm/v1/engine/async_llm.py:138: error: Incompatible types in assignment (expression has type "type[UniprocExecutor]", variable has type "type[MultiprocExecutor]")  [assignment]
Found 7 errors in 5 files (checked 33 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: fix overloaded getitem/setitem in ConstantList

Fixes:

vllm/v1/utils.py:37: error: Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader  [misc]
vllm/v1/utils.py:48: error: Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader  [misc]
...
Found 52 errors in 12 files (checked 30 source files)

Now:

Found 46 errors in 11 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: fix already defined error in Request

Fixes:

vllm/v1/request.py:58: error: Attribute "mm_inputs" already defined on line 56  [no-redef]

Now:

Found 45 errors in 10 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: add type hints to EngineCoreProc queues

Fixes:

vllm/v1/engine/core.py:144: error: Need type annotation for "input_queue"  [var-annotated]
vllm/v1/engine/core.py:145: error: Need type annotation for "output_queue"  [var-annotated]

Now:

Found 43 errors in 10 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: ignore not defined errors with zmq.Socket/Context

Same workaround already used in v0.

Fixes:

vllm/v1/engine/core.py:158: error: Name "zmq.Socket" is not defined  [name-defined]
vllm/v1/engine/core.py:161: error: Module has no attribute "Context"  [attr-defined]
vllm/v1/engine/core_client.py:130: error: Module has no attribute "Context"  [attr-defined]

Now:

Found 40 errors in 9 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: make Processor.process_inputs(arrival_time) optional

Fixes:

vllm/v1/engine/llm_engine.py:137: error: Argument 4 to "process_inputs" of "Processor" has incompatible type "Optional[float]"; expected "float"  [arg-type]
vllm/v1/engine/async_llm.py:147: error: Argument 4 to "process_inputs" of "Processor" has incompatible type "Optional[float]"; expected "float"  [arg-type]

Now:

Found 34 errors in 8 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: add profile_async() to EngineCoreClient/AsyncMPClient

Fixes:

vllm/v1/engine/llm_engine.py:167: error: Value of type "Coroutine[Any, Any, None]" must be used  [unused-coroutine]
vllm/v1/engine/llm_engine.py:167: note: Are you missing an await?
vllm/v1/engine/llm_engine.py:170: error: Value of type "Coroutine[Any, Any, None]" must be used  [unused-coroutine]
vllm/v1/engine/llm_engine.py:170: note: Are you missing an await?

Now:

Found 32 errors in 7 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: add type hint to AsyncLLM.output_handler

Fixes:

vllm/v1/engine/async_llm.py:193: error: Incompatible types in assignment (expression has type "Task[Any]", variable has type "None")  [assignment]

Now:

Found 31 errors in 7 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: fix type error with AsyncLLM.dead_error

Fixes:

vllm/v1/engine/async_llm.py:368: error: Incompatible return value type (got "type[Exception]", expected "BaseException")  [return-value]

Now:

Found 30 errors in 6 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: make FreeKVCacheBlockQueue free_list_head/tail optional

Fixes:

vllm/v1/core/kv_cache_utils.py:111: error: Incompatible types in assignment (expression has type "Optional[KVCacheBlock]", variable has type "KVCacheBlock")  [assignment]
vllm/v1/core/kv_cache_utils.py:114: error: Incompatible types in assignment (expression has type "Optional[KVCacheBlock]", variable has type "KVCacheBlock")  [assignment]
vllm/v1/core/kv_cache_utils.py:150: error: Incompatible types in assignment (expression has type "Optional[KVCacheBlock]", variable has type "KVCacheBlock")  [assignment]

Now:

Found 27 errors in 6 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: fix reversed type hint in KVCacheManager

Fixes:

vllm/v1/core/kv_cache_manager.py:268: error: Incompatible types in assignment (expression has type "reversed[KVCacheBlock]", variable has type "list[KVCacheBlock]")  [assignment]

Now:

Found 22 errors in 4 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: add new_blocks assertion in scheduler

Fixes:

vllm/v1/core/scheduler.py:157: error: Item "None" of "Optional[list[KVCacheBlock]]" has no attribute "__iter__" (not iterable)  [union-attr]

Now:

Found 21 errors in 3 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: fix stop_token type hint in add_tokens()

Fixes:

vllm/v1/engine/detokenizer.py:258: error: Argument "stop_reason" to "add_tokens" of "IncrementalDetokenizer" has incompatible type "Union[int, str, None]"; expected "Optional[str]"  [arg-type]

Now:

Found 20 errors in 2 files (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: copy flash atten forward pass assertion from v0

Fixes:

vllm/v1/attention/backends/flash_attn.py:162: error: Value of type "Optional[Any]" is not indexable  [index]

Now:

Found 19 errors in 1 file (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: don't reuse req_data with different types

Fixes:

vllm/v1/worker/gpu_model_runner.py:164: error: Incompatible types in assignment (expression has type "NewRequestData", variable has type "RunningRequestData")  [assignment]
vllm/v1/worker/gpu_model_runner.py:166: error: "RunningRequestData" has no attribute "sampling_params"  [attr-defined]
vllm/v1/worker/gpu_model_runner.py:175: error: "RunningRequestData" has no attribute "prompt_token_ids"  [attr-defined]
vllm/v1/worker/gpu_model_runner.py:176: error: "RunningRequestData" has no attribute "prompt"  [attr-defined]
vllm/v1/worker/gpu_model_runner.py:177: error: "RunningRequestData" has no attribute "mm_inputs"  [attr-defined]
vllm/v1/worker/gpu_model_runner.py:178: error: "RunningRequestData" has no attribute "mm_positions"  [attr-defined]
vllm/v1/worker/gpu_model_runner.py:181: error: "RunningRequestData" has no attribute "block_ids"; maybe "new_block_ids"?  [attr-defined]
vllm/v1/worker/gpu_model_runner.py:188: error: Incompatible types in assignment (expression has type "ResumedRequestData", variable has type "RunningRequestData")  [assignment]
vllm/v1/worker/gpu_model_runner.py:192: error: "RunningRequestData" has no attribute "block_ids"; maybe "new_block_ids"?  [attr-defined]

Now:

Found 10 errors in 1 file (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: add req_id is not None assertion in GPUModelRunner

Fixes:

vllm/v1/worker/gpu_model_runner.py:230: error: Invalid index type "Optional[str]" for "dict[str, int]"; expected type "str"  [index]

Now:

Found 9 errors in 1 file (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: fix req_input_ids type

Fixes:

vllm/v1/worker/gpu_model_runner.py:365: error: Argument 1 to "append" of "list" has incompatible type "tuple[str, int]"; expected "tuple[int, int]"  [arg-type]
vllm/v1/worker/gpu_model_runner.py:381: error: Incompatible types in assignment (expression has type "int", variable has type "str")  [assignment]

Now:

Found 6 errors in 1 file (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: add some of req_id is not None assertions

Fixes:

vllm/v1/worker/gpu_model_runner.py:394: error: Invalid index type "Optional[str]" for "dict[str, int]"; expected type "str"  [index]
vllm/v1/worker/gpu_model_runner.py:395: error: Invalid index type "Optional[str]" for "dict[str, CachedRequestState]"; expected type "str"  [index]
vllm/v1/worker/gpu_model_runner.py:489: error: Invalid index type "Optional[str]" for "dict[str, CachedRequestState]"; expected type "str"  [index]
vllm/v1/worker/gpu_model_runner.py:491: error: Invalid index type "Optional[str]" for "dict[str, int]"; expected type "str"  [index]
vllm/v1/worker/gpu_model_runner.py:810: error: Invalid index type "Optional[str]" for "dict[str, int]"; expected type "str"  [index]

Now:

Found 1 error in 1 file (checked 30 source files)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: tell mypy to ignore erros with msgspec.Struct kwargs

As per the similar v0 code.

Fixes:

vllm/v1/engine/__init__.py:46: error: Unexpected keyword argument "array_like" for "__init_subclass__" of "object"  [call-arg]
vllm/v1/engine/__init__.py:46: error: Unexpected keyword argument "omit_defaults" for "__init_subclass__" of "object"  [call-arg]
vllm/v1/engine/__init__.py:46: error: Unexpected keyword argument "gc" for "__init_subclass__" of "object"  [call-arg]
vllm/v1/engine/__init__.py:58: error: Unexpected keyword argument "array_like" for "__init_subclass__" of "object"  [call-arg]
vllm/v1/engine/__init__.py:58: error: Unexpected keyword argument "omit_defaults" for "__init_subclass__" of "object"  [call-arg]
vllm/v1/engine/__init__.py:58: error: Unexpected keyword argument "gc" for "__init_subclass__" of "object"  [call-arg]

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented on vllm/v1/engine/core_client.py:
```diff
@@ -71,6 +74,9 @@ async def get_output_async(self) -> List[EngineCoreOutput]:
     async def add_request_async(self, request: EngineCoreRequest) -> None:
         raise NotImplementedError
 
+    async def profile_async(self, is_start=True) -> None:
```
Done across the board, thanks

markmc commented on vllm/v1/core/kv_cache_utils.py:
```diff
@@ -152,7 +152,7 @@ def get_all_free_blocks(self) -> List[KVCacheBlock]:
 
 def hash_block_tokens(parent_block_hash: Optional[int],
-                      curr_block_token_ids: Tuple[int]) -> BlockHashType:
+                      curr_block_token_ids: List[int]) -> BlockHashType:
```
It's actually the revers AFAICT - in hash_request_tokens() we were converting the token IDs from their natural list[int] form to tuple[int]

```
block_token_ids = tuple(token_ids[start:end])
```

I think just removing that conversion to tuple makes more sense?

markmc replied:
ok, to avoid the conversion from ConstantList to list elsewhere, I've added:

```
TokenIdsType = Union[List[int], ConstantList[int]]
BlockHashType = Tuple[int, TokenIdsType]
```

Collection didn't work, because it needs to be indexable/sliceable:

```
    for start in range(0, len(token_ids), block_size):
        end = start + block_size
	block_token_ids = token_ids[start:end]
```

DarkLight1337 replied:
How about `Sequence[int]` then?

markmc replied:
Good idea, done now :+1:

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -223,9 +223,10 @@ def _prepare_inputs(self, scheduler_output: "SchedulerOutput"):
         # Get the number of scheduled tokens for each request.
         # TODO: The Python loop can be slow. Optimize.
-        num_scheduled_tokens = []
+        num_scheduled_tokens = np.empty(0, dtype=np.int32)
```
Can we keep this a list but just rename it? I'd like to avoid using `append` for numpy arrays.

markmc replied:
I'll revert this, I think #11111 removed the original error. Great catch :+1:

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -509,8 +512,14 @@ def execute_model(
+        assert all(
+            req_id is not None for req_id in
+            self.input_batch.req_ids[:num_reqs]), "req_ids contains None"
+        req_ids = cast(List[str], self.input_batch.req_ids[:num_reqs])
```
QQ: why is this line needed?

markmc replied:
This is the error:

```
vllm/v1/worker/gpu_model_runner.py:546: error: Argument "req_ids" to "ModelRunnerOutput" has incompatible type "list[Optional[str]]"; expected "list[str]"  [arg-type]
```

`input_batch.req_ids` can have `None` elements:

```
self.req_ids: List[Optional[str]] = [None] * max_num_reqs
```

but `ModelRunnerOutput.req_ids` can't:

```
class ModelRunnerOutput:

    # [num_reqs]                                                                                                                                                                                                                                                                  
    req_ids: List[str]
```

AFAIR `input_batch.num_reqs` tells us how many are non-None?

WoosukKwon replied:
Got it. Thanks!

markmc committed:
v1: add type hints to LRUDictCache

Fixes:

vllm/v1/utils.py:110: error: Need type annotation for "cache"  [var-annotated]

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: remove return type from cache_hit_ratio()

Fixes:

vllm/v1/engine/mm_input_mapper.py:52: error: Missing return statement  [return]

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: mm_hashes type hints

Fixes:

vllm/v1/engine/mm_input_mapper.py:76: error: Argument 1 to "len" has incompatible type "Optional[list[str]]"; expected "Sized"  [arg-type]
vllm/v1/engine/mm_input_mapper.py:77: error: Argument 1 to "len" has incompatible type "Optional[list[str]]"; expected "Sized"  [arg-type]
vllm/v1/engine/mm_input_mapper.py:82: error: Need type annotation for "ret_hashes"  [var-annotated]
vllm/v1/engine/mm_input_mapper.py:92: error: Value of type "Optional[list[str]]" is not indexable  [index]
vllm/v1/engine/mm_input_mapper.py:110: error: Argument 1 to "put" of "LRUDictCache" has incompatible type "Optional[str]"; expected "str"  [arg-type]
vllm/v1/engine/mm_input_mapper.py:116: error: Item "None" of "Optional[list[str]]" has no attribute "append"  [union-attr]
vllm/v1/engine/mm_input_mapper.py:116: error: Argument 1 to "append" of "list" has incompatible type "Optional[str]"; expected "str"  [arg-type]
vllm/v1/engine/mm_input_mapper.py:119: error: Incompatible return value type (got "tuple[list[Any], Optional[list[str]]]", expected "list[Any]")  [return-value]
vllm/v1/engine/core.py:99: error: Argument 1 to "process_inputs" of "MMInputMapperServer" has incompatible type "Optional[list[Optional[Any]]]"; expected "list[Optional[Any]]"  [arg-type]

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: add another MultiprocExecutor.workers assertion

Fixes:

vllm/v1/executor/multiproc_executor.py:190: error: Item "None" of "Optional[list[WorkerProcHandle]]" has no attribute "__iter__" (not iterable)  [union-attr]

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: make mm_hashes List[str] not List[Optional[str]]

Seems this will never contain None values:

Fixes:

vllm/v1/engine/processor.py:143: error: Argument 5 to "EngineCoreRequest" has incompatible type "Optional[list[str]]"; expected "Optional[list[Optional[str]]]"  [arg-type]

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
@DarkLight1337 @WoosukKwon note that I've rebased and fixed a bunch of new issues. Those new fixes are in the 5 latest commits

markmc commented:
rebased again to fix conflicts

@comaniac would you mind reviewing commit 9190f342ffc4edfae851c4f0216ffc1231cde4ec ?

For reference, in #9972 you said:

> Enhance the block hash type from int to Tuple[int, Tuple[int]], which is (hash_value, (toke_ids,)). This guarantees no hash conflicts

Pasting the commit message:

The current block hash format is:

```
  (hash((parent_hash, *token_id_list)), token_id_list)
```

i.e. it is a tuple with two parts:

1. A hash of a tuple combining the parent hash with token ids
2. A tuple containing the same token ids

and this (hash, token_ids) tuple itself must be hashable in order to be a dictionary key.

This is suspicious - why include the same token IDs twice in the same hash?

And then this code:

```
def hash_request_tokens(block_size: int,
                        token_ids: List[int]) -> List[BlockHashType]:
    ...
    ret = []
    parent_block_hash = None
    for ...:
        tuple(token_ids[start:end])
        ...
        block_hash = hash_block_tokens(parent_block_hash, block_token_ids)
        parent_block_hash = block_hash
    ...
```

looks wrong because hash_block_tokens() expects an integer hash for the parent_block_hash argument:

```
def hash_block_tokens(parent_block_hash: Optional[int],
                      curr_block_token_ids: Tuple[int]) -> BlockHashType:
```

It appears like it would be perfectly sufficient to use the integer hash of the `(parent_hash + token_ids)` tuple.

If we do this, then the curr_block_token_ids argument no longer needs to be a tuple - but it does need to be a sliceable sequence, so let's make sure ConstantList satisfies that requirement.

comaniac commented:
> This is suspicious - why include the same token IDs twice in the same hash?

This is to avoid potential hash collusion, because Python dict guarantees no collusion if the dict key is tuple.
But yeah the type is incorrect as you said. I'll submit a fix PR in minutes.

markmc committed:
v1: fix block_hash mypy errors

BlockHashType token IDs should be Tuple[int, ...] - it's not just
one int.

Also, pass the token IDs as a Sequence[int] and convert to a tuple
only when we need it in hash_block_tokens().

Make ConstantList a Sequence subclass to allow this.

Fixes:

vllm/v1/core/kv_cache_utils.py:206: error: Argument 2 to "hash_block_tokens" has incompatible type "tuple[int, ...]"; expected "tuple[int]"  [arg-type]
vllm/v1/core/kv_cache_manager.py:89: error: Argument 2 to "hash_request_tokens" has incompatible type "ConstantList[int]"; expected "list[int]"  [arg-type]
vllm/v1/core/kv_cache_manager.py:401: error: Argument 2 to "hash_block_tokens" has incompatible type "tuple[int, ...]"; expected "tuple[int]"  [arg-type]

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
> > This is suspicious - why include the same token IDs twice in the same hash?
> 
> This is to avoid potential hash collision, because Python dict guarantees no collision if the dict key is tuple. But yeah the type is incorrect as you said. I'll submit a fix PR in minutes.

Thanks for the fix! Rebased again to resolve merge conflicts and fixed a few mypy errors with this prefix caching code.

markmc commented:
Moving to draft while I debug the test failures

markmc committed:
v1: ignore duplicate keyword argument mypy errors

Not very satisfactory, but other attempts caused regressions.

Fixes:

vllm/v1/engine/core_client.py:158: error: "make_engine_core_process" of "EngineCoreProc" gets multiple values for keyword argument "input_path"  [misc]
vllm/v1/engine/core_client.py:158: error: "make_engine_core_process" of "EngineCoreProc" gets multiple values for keyword argument "output_path"  [misc]
vllm/v1/engine/core_client.py:158: error: "make_engine_core_process" of "EngineCoreProc" gets multiple values for keyword argument "ready_path"  [misc]

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

tlrmchlsmth commented on vllm/v1/executor/multiproc_executor.py:
The asserts on the workers here are pretty nasty.

Setting the workers to `None` during `shutdown` was a bit of paranoia on my end to make sure everything was cleaned up.

I suggest changing shutdown to the following, and then we can ditch the asserts
```
    def shutdown(self):
        """Properly shut down the executor and its workers"""
        if getattr(self, 'shutting_down', False):
            self.shutting_down = True
            for w in self.workers:  #TODO: not sure if needed
                w.worker_response_mq = None
            self._ensure_worker_termination()

        self.rpc_broadcast_mq = None
```

markmc replied:
Makes sense, thanks. Done now.

That's all 3 of my "dubious changes" sorted! :+1:

markmc committed:
v1: Make MultiprocExecutor.workers non-optional

Suggestion from @tlrmchlsmth.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

tlrmchlsmth commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -103,7 +104,7 @@ def collective_rpc(self,
                        timeout: Optional[float] = None,
                        args: Tuple = (),
-                       kwargs: Optional[Dict] = None) -> []:
+                       kwargs: Optional[Dict] = None) -> Any:
```
```suggestion
                       kwargs: Optional[Dict] = None) -> List[Any]:
```

markmc replied:
Thanks, good point

tlrmchlsmth commented on vllm/v1/executor/uniproc_executor.py:
Similar to `multiproc_executor`, we don't need to set worker to None in `shutdown`, which will let us remove the asserts

markmc replied:
Doh. Of course. Done now.

After a rebase, another one of these crept in, but doesn't require asserts so I didn't touch it. See https://github.com/vllm-project/vllm/pull/11074/files#r1884597768

tlrmchlsmth reviewed (approved):
Thanks so much for this one!

markmc committed:
v1: Make UniprocExecutor.worker non-optional

Suggestion from @tlrmchlsmth.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: fix MultiprocExecutor.collective_rpc() return type

Suggestion by @tlrmchlsmth.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
v1: MPClient.proc-handle can be None

Fixes:

vllm/v1/engine/core_client.py:196: error: Incompatible types in assignment (expression has type "None", variable has type "EngineCoreProcHandle")  [assignment]

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

WoosukKwon commented:
@markmc Finally merged! Thanks for your patience and great work! Really appreciate it.
